core/src/main/kotlin/org/evomaster/core/search/impact/ImpactUtils.kt

```diff
@@ -0,0 +1,179 @@
+            }
+            individual.seeActions().find { a-> a.seeGenes().contains(gene) }?.let {
+                return generateGeneId(it, gene)
+            }
+            return generateGeneId(gene)
+        }
+
+        fun extractMutatedGeneWithContext(mutatedGenes : MutableList<Gene>, individual: Individual, previousIndividual: Individual) : Map<String, MutableList<MutatedGeneWithContext>>{
+            val mutatedGenesWithContext = mutableMapOf<String, MutableList<MutatedGeneWithContext>>()
+
+            if (individual.seeActions().isEmpty()){
+                //throw IllegalArgumentException("do not support to extract contexts of mutated genes for an individual which does not have any action, i.e., seeAction() is empty.")
+                individual.seeGenes().filter { mutatedGenes.contains(it) }.forEach { g->
+                    val id = generateGeneId(individual, g)
+                    val contexts = mutatedGenesWithContext.getOrPut(id){ mutableListOf()}
+                    val previous = findGeneById(previousIndividual, id)?: throw IllegalArgumentException("mismatched previous individual")
+                    contexts.add(MutatedGeneWithContext(g, previous = previous))
+                }
+            }else{
+                individual.seeActions().forEachIndexed { index, action ->
+                    action.seeGenes().filter { mutatedGenes.contains(it) }.forEach { g->
+                        val id = generateGeneId(action, g)
+                        val contexts = mutatedGenesWithContext.getOrPut(id){ mutableListOf()}
+                        val previous = findGeneById(previousIndividual, id, action.getName(), index, false)?: throw IllegalArgumentException("mismatched previous individual")
+                        contexts.add(MutatedGeneWithContext(g, action.getName(), index, previous))
+                    }
+                }
+            }
+
+            assert(mutatedGenesWithContext.values.sumBy { it.size } == mutatedGenes.size)
+            return mutatedGenesWithContext
+        }
+
+        fun extractMutatedDbGeneWithContext(mutatedGenes : MutableList<Gene>, individual: Individual, previousIndividual: Individual) : Map<String, MutableList<MutatedGeneWithContext>>{
+            val mutatedGenesWithContext = mutableMapOf<String, MutableList<MutatedGeneWithContext>>()
+
+            individual.seeInitializingActions().forEachIndexed { index, action ->
+                action.seeGenes().filter { mutatedGenes.contains(it) }.forEach { g->
+                    val id = generateGeneId(action, g)
+                    val contexts = mutatedGenesWithContext.getOrPut(id){ mutableListOf()}
+                    val previous = findGeneById(previousIndividual, id, action.getName(), index, true)
+                    contexts.add(MutatedGeneWithContext(g, action.getName(), index, previous))
+                }
+            }
+
+            assert(mutatedGenesWithContext.values.sumBy { it.size } == mutatedGenes.size)
```

Comment:
as u are iterating over a collection to sum values, it would be better to put this in a `Lazy.assert`.